Add zmon built in command for getting data from s3 buckets

lorenzhawkes opened the pull request:
Add a new s3_bucket() command that allows the user to fetch data from s3 buckets. 
The contents of the s3 bucket can be retrieved as is, or as JSON.

hjacobs commented on zmon_worker_monitor/builtins/plugins/s3_bucket.py:
```diff
@@ -0,0 +1,54 @@
+        data = StringIO.StringIO()
+        self.__client.download_fileobj(bucket_name, key, data)        
+        data.seek(0)
+        result = json.load(data)
```
always use `safe_load`!

mohabusama replied:
I believe ``safe_load`` is in yaml and not json. But ``try...finally`` would be safer to call ``data.close``

hjacobs commented:
I would rename the wrapper to `S3Wrapper` and only have `get_object` with another helper object to have a simpler way of invocation and to match the existing `requests` library style:

```
s3().get_object('my-bucket', 'foo/bar').text()
s3().get_object('my-bucket', 'foo/bar').json()

mohabusama commented on zmon_worker_monitor/builtins/plugins/s3_bucket.worker_plugin:
```diff
@@ -0,0 +1,8 @@
+
+[Documentation]
+Version = 0.1
+Website = https://github.com/zalando/zmon-worker
```
`zalando-zmon/zmon-worker`

lorenzhawkes commented:
:+1:

hjacobs commented on zmon_worker_monitor/builtins/plugins/s3.py:
```diff
@@ -0,0 +1,59 @@
+logging.getLogger('botocore').setLevel(logging.WARN)
+
+
+class S3BucketWrapperFactory(IFunctionFactoryPlugin):
```
rename to `S3WrapperFactory`?

hjacobs commented on zmon_worker_monitor/builtins/plugins/s3.py:
```diff
@@ -0,0 +1,59 @@
+        return S3Object(result)
+
+
+class S3Object(object):
```
rename to `S3Wrapper`?

mohabusama commented:
Just couple of concerns:

1. What will happen with extremely large files? In the worker we define a hard limit on result size, but with s3 buckets we could load really large files before size limit assertion.
2. I imagine some would need to check if file exists or just its metadata (e.g size) without the need to download the object, so won't it be more convenient (+ efficient) to add ``.size()`` and ``.exists()``?

lorenzhawkes commented:
I've added the metadata methods for an S3Object and to get them from the S3Bucket without having to fetch the entrire S3Object first of all.

I've updated the unit tests so they behave correctly under the error conditions I've observed when testing locally.

hjacobs commented:
@mohabusama can you review & approve if you think it's ok?

Also important: documentation!

lorenzhawkes commented:
Added docs for everything

hjacobs commented:
@lorenzhawkes thanks. Sorry, I meant user docs on https://docs.zmon.io/: https://github.com/zalando-zmon/zmon-docs

mohabusama commented on zmon_worker_monitor/builtins/plugins/s3.py:
```diff
@@ -0,0 +1,138 @@
+    return r.json()['region']
+
+
+class S3Bucket(object):
```
Please change to ``S3Wrapper`` for consistency with other plugins.

mohabusama commented:
Just a general remark on the interface, a rather more consistent one could be:

```python
s3('my-bucket', 'my-key').size()
s3('my-bucket', 'my-key').exists()
s3('my-bucket', 'my-key').json()
s3('my-bucket', 'my-key').text()
```
so in a way, we would abstract ``get_object`` and ``get_object_metadata`` from the user.

However, we can go with current implementation and depend on proper documentation on https://docs.zmon.io

@hjacobs, @lorenzhawkes what do you think?

hjacobs commented:
@mohabusama I think the current behavior is fine --- it's not consistent with HttpWrapper, but it makes sense here as we might want to have a call like `list_buckets` in the future.
I just kind of dislike the extra `get_object_metadata` (but I don't have a better suggestion now):

```
s3().get_object('my-bucket', 'my-key').json()
s3().list_buckets() # suggestion, maybe "ls" with pattern is better

mohabusama commented:
👍

hjacobs commented:
:+1:
